cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/OperationsDashboardHttpHandler.java

```diff
@@ -77,36 +109,104 @@ public void readDashboardDetail(FullHttpRequest request, HttpResponder responder
     }
 
     Set<NamespaceId> namespaceIds = namespaces.stream().map(NamespaceId::new).collect(Collectors.toSet());
+    long endTimeSecs = startTimeSecs + durationTimeSecs;
     Map<ProgramRunId, RunRecordMeta> historicalRuns =
       // TODO: [CDAP-13352] Currently, to get active program runs within a time range,
       // a full table scan is required in AppMetaStore. Performance improvement will be done.
-      store.getHistoricalRuns(namespaceIds, startTimeSecs, startTimeSecs + durationTimeSecs, Integer.MAX_VALUE);
+      store.getHistoricalRuns(namespaceIds, startTimeSecs, endTimeSecs, Integer.MAX_VALUE);
     // get historical runs within the query time range
     Stream<DashboardProgramRunRecord> historicalRecords =
       historicalRuns.values().stream().map(OperationsDashboardHttpHandler::runRecordToDashboardRecord);
     // get active runs with start time earlier than the end of query time range
     Stream<DashboardProgramRunRecord> activeRecords =
-      store.getActiveRuns(namespaceIds, run -> run.getStartTs() < startTimeSecs + durationTimeSecs).values().stream()
-      .map(OperationsDashboardHttpHandler::runRecordToDashboardRecord);
-    responder.sendJson(HttpResponseStatus.OK,
-                       // combine historical runs and active runs
-                       GSON.toJson(Stream.concat(historicalRecords, activeRecords).collect(Collectors.toSet())));
+      store.getActiveRuns(namespaceIds, run -> run.getStartTs() < endTimeSecs).values().stream()
+        .map(OperationsDashboardHttpHandler::runRecordToDashboardRecord);
+    List<DashboardProgramRunRecord> result = new ArrayList<>();
+    // add combined historical runs and active runs to the result
+    result.addAll(Stream.concat(historicalRecords, activeRecords).collect(Collectors.toList()));
+    // if the end time is in the future, also add scheduled program runs to the result
+    if (endTimeSecs > TimeUnit.MICROSECONDS.toSeconds(System.currentTimeMillis())) {
+      // get enabled time schedules from all given namespaces
+      getEnabledTimeSchedules(namespaceIds)
+        // for each schedule, add all the scheduled runs within given the time range to the result
+        .forEach(schedule -> {
+          try {
+            result.addAll(getScheduledDashboardRecords(schedule, startTimeSecs, endTimeSecs));
+          } catch (Exception e) {
+            LOG.error("Failed to get scheduled program runs for schedule {}", schedule, e);
+          }
+        });
+    }
```

Comment:
extract this method into smaller methods, `getRuns` and `getFutureScheduledRuns`, better to have smaller methods than a single giant method.